userhelper.8
.\" Copyright (C) 1997,1999 Red Hat Software, Inc.
.\"
.\" This is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH USERHELPER 8 "23 August 2000" "Red Hat, Inc."
.SH NAME
userhelper \- A helper interface to pam.
.SH SYNOPSIS
.B userhelper
[\ -t\ ] [\ -d in,out,err\ ] [\ -w\ prog\ args\ ] [\ -c\ ] [\ \-f\ full-name\ ] [\ \-o\ office\ ] [\ \-p\ office-phone\ ] [\ \-h\ home-phone\ ] [\ \-s\ shell\ ] [\ username \]

.SH DESCRIPTION
NOTE this program is NOT intended to be run interactively.  If you
want to change this information on the command line use
.IR passwd (1),
.IR chfn (1), 
or
.IR chsh (1).
.PP
This program provides a basic interface to change a user's password, gecos
information, and shell.  The main differnce between this program and
its traditional equivilents is that prompts come out on standard out
to make it easy for a GUI program to interface it as a child process.
.PP
The output is in the form of:
.TP
<number> <string>
.PP
Where the number is the type of prompt returned from pamlib, and the
string is the prompt to give the user.
.PP
The prompt numbers are as follows:
.TP
1
Prompt with visible input.
.TP
2
Prompt with invisible input.
.TP
3
Informational message.
.TP
4
Error message.
.TP
5
Count of messages sent in this block so far.
.TP
6
The name of the service being used.
.TP
7
Whether or not the command will be executed as the user if authentication fails.
.TP
8
The name of the user being authenticated.

.SH OPTIONS
.TP
.I -d
A comma-separated list of three descriptors which need to be reattached to
the standard handles before the command is executed; only used with -w.
.TP
.I -t
Text mode authentication instead of the numbered message types
just described; only used with -w.
.TP
.I -w
Specify a program name and arguments to be passed through.  userhelper
will look up in the file /etc/security/console.apps/\fIprogramname\fP
the username to authenticate stored as USER=\fIvalue\fP (\fIvalue\fP
is normally "root" or the magic token "<user>" which indicates
to authenticate the current user) and the program to run stored as
PROGRAM=\fIvalue\fP (\fIvalue\fP \fBmust\fP be an absolute path or it
will be ignored).  If USER is not specified, the current user is used,
and if PROGRAM is not specified, userhelper looks for \fIprogramname\fP
in /sbin and /usr/sbin/.  userhelper will then authenticate the user via
pam and then run the named program.  If SESSION=true is specified, then
userhelper establishes a pam session and forks and execs, then waits
around for the child to exit; otherwise, it simply execs the child.
userhelper is normally called from consolehelper via a symlink.
If FALLBACK=true is specified, the program will run as user if
authentication fails.  Authentication will be retried twice (three
times total) unless something other than RETRY=2 is specified.
.TP
.I -c 
Change the current user's password.  Note that this option cannot be
used with any of the other options.  This is due to the limitation in
the interface to pamlib.
.TP
.I -f
Specify a new Full Name.
.TP
.I -o
Specify a new Office.
.TP
.I -p
Specify a new Office Phone.
.TP
.I -h
Specify a new Home Phone.
.TP
.I -s
Specify a new shell.
.SH EXIT STATUS
A non-zero exit status indicates an error occured.  Those errors are:
.TP
1
The authentication passwords was incorrect.
.TP
2
One or more of the GECOS fields is invalid.  This occurs when there is
a colon supplied in one of the fields.
.TP
3
Password resetting error.
.TP
4
Some system files are locked.
.TP
5
User unknown.
.TP
6
Insufficent rights.
.TP
7
Invalid call to this program.
.TP
8
The shell provided is not valid (i.e., does not exist in /etc/shells).
.TP
9
Ran out of memory.
.TP
10
Could not find the program.
.TP
11
exec failed even though program exists.
.TP
255
Unknown error.
.SH FILES
.TP 25
.I /etc/passwd
The gecos and shell information is stored in this file.
.TP 25
.I /etc/shells
This file is checked to see if the new shell supplied is valid.
.SH "SEE ALSO"
.IR userpasswd (1),
.IR userinfo (1),
.IR consolehelper (8),
.IR chfn (1),
.IR chsh (1),
.IR passwd (5)
.SH AUTHOR
Otto Hammersmith <otto@redhat.com>
.br
Michael K. Johnson <johnsonm@redhat.com>


NEWS
1.99
* Support multi-line prompts.

1.98.1
* Update translations

1.98
* Support dialogs with no text entries.
* Left-justify prompts.
* Remove /usr/X11R6/bin from default path.

1.97
* Fix display of '_' in prompts.

1.96.1
* Update translations

1.96
* Remove code that overrides SELinux contexts of processes started by
  userhelper.

1.95
* New home page at https://fedoraproject.org/usermode/ .
* Correctly preserve exit code when SESSION=yes
* Fix minor errors in the .desktop files

1.94
* Add support for including files from wrapper configuration files.  Original
  patch by Carlo de Wolf <cdewolf@redhat.com>.

1.93.1
* Allow building with startup notification support without libwnck
* Update translations

1.93
* Fix handling of PAM conversations with no questions

1.92
* Fix userhelper hangs in non-UTF8 locales.
* Fix unexpanded {prefix} strings embedded in executables if some directory
  variables were not explicitly passed to configure.
* Add menu categories to desktop files
* Minor clean-ups and optimizations.

1.91.2
* Show the user we're authenticating as in an i18n-safe way.

1.91.1
* Update translations

1.91
* Preserve environment variables in consolehelper specified in the service
  config file.
* Fix handling of empty pw_shell.
* Minor clean-ups and optimizations.

1.90
* Support LABEL= and UUID= in usermount and userformat.
* Fix an use of PAM data after free.
* Add a dialog caption when changing user password.
* Use less memory in pam-panel-icon until the icon is actually displayed.
* Don't use deprecated GTK+ widgets.
* Other minor bug fixes, clean-ups and optimizations.

1.89
* Preserve the exit status of processes executed by consolehelper.


userhelper.c
/* -*-Mode: c-*- */
/* Copyright (C) 1997 Red Hat Software, Inc.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
/* 
 * UserTool suid helper program
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <unistd.h>
#include <pwd.h>
#include <sys/types.h>

#include <security/pam_appl.h>
#include <security/pam_misc.h>

#include <pwdb/pwdb_public.h>

#include "userhelper.h"

/* Total GECOS field length... is this enough ? */
#define GECOS_LENGTH		80

#ifndef TRUE
#define TRUE 1
#endif
#ifndef FALSE
#define FALSE 0
#endif


/* ------ some static data objects ------- */

static char *full_name	= NULL; /* full user name */
static char *office	= NULL; /* office */
static char *office_ph	= NULL;	/* office phone */
static char *home_ph	= NULL;	/* home phone */
static char *user_name	= NULL; /* the account name */
static char *shell_path = NULL; /* shell path */

/* command line flags */
static int 	f_flg = 0; 	/* -f flag = change full name */
static int	o_flg = 0;	/* -o flag = change office name */
static int	p_flg = 0;	/* -p flag = change office phone */
static int 	h_flg = 0;	/* -h flag = change home phone number */
static int 	c_flg = 0;	/* -c flag = change password */
static int 	s_flg = 0;	/* -s flag = change shell */
static int 	t_flg = 0;	/* -t flag = direct text-mode -- exec'ed */
static int 	w_flg = 0;	/* -w flag = act as a wrapper for next args */

/*
 * A handy fail exit function we can call from man places
 */
static int fail_error(int retval)
{
  /* this is a temporary kludge.. will be fixed shortly. */
    if(retval == ERR_SHELL_INVALID)
        exit(ERR_SHELL_INVALID);	  

    if (retval != PAM_SUCCESS) {
	switch(retval) {
	    case PAM_AUTH_ERR:
	    case PAM_PERM_DENIED:
		exit (ERR_PASSWD_INVALID);
	    case PAM_AUTHTOK_LOCK_BUSY:
		exit (ERR_LOCKS);
	    case PAM_CRED_INSUFFICIENT:
	    case PAM_AUTHINFO_UNAVAIL:
		exit (ERR_NO_RIGHTS);
	    case PAM_ABORT:
	    default:
		exit(ERR_UNK_ERROR);
	}
    }
    exit (0);
}

/*
 * Read a string from stdin, returns a malloced memory to it
 */
static char *read_string(void)
{
    char *buffer = NULL;
    char *check = NULL;
    int slen = 0;
    
    buffer = (char *) malloc(BUFSIZ);
    if (buffer == NULL)
	return NULL;
    
    check = fgets(buffer, BUFSIZ, stdin);
    if (check != buffer)
	return NULL;
    slen = strlen(buffer);
    if (buffer[slen-1] == '\n')
	buffer[slen-1] = '\0';
    return buffer;
}

/*
 * Conversation function for the boring change password stuff
 */
static int conv_func(int num_msg, const struct pam_message **msg,
		     struct pam_response **resp, void *appdata_ptr)
{
    int 	count = 0;
    struct pam_response *reply = NULL;

    reply = (struct pam_response *)
	calloc(num_msg, sizeof(struct pam_response));
    if (reply == NULL)
	return PAM_CONV_ERR;
    
    /*
     * We do first a pass on all items and output them;
     * then we do a second pass and read what we have to read
     * from stdin
     */
    for (count = 0; count < num_msg; count++) {
	switch (msg[count]->msg_style) {
	    case PAM_PROMPT_ECHO_ON:
		printf("%d %s\n", UH_ECHO_ON_PROMPT, msg[count]->msg);
		break;
	    case PAM_PROMPT_ECHO_OFF:
		printf("%d %s\n", UH_ECHO_OFF_PROMPT, msg[count]->msg);
		break;
	    case PAM_TEXT_INFO:
		printf("%d %s\n", UH_INFO_MSG, msg[count]->msg);
		break;
	    case PAM_ERROR_MSG:
		printf("%d %s\n", UH_ERROR_MSG, msg[count]->msg);
		break;
	    default:
		printf("0 %s\n", msg[count]->msg);
	}
    }

    /* tell the other side how many messages we sent and how many
     * responses we expect (ignoring messages, which we fudge here).
     */
    printf("%d %d", UH_EXPECT_RESP, num_msg);

    /* now the second pass */
    for (count = 0; count < num_msg; count++) {
	switch (msg[count]->msg_style) {
	    case PAM_PROMPT_ECHO_ON:
		reply[count].resp_retcode = PAM_SUCCESS;
		reply[count].resp = read_string();
		/* PAM frees resp */
		break;
	    case PAM_PROMPT_ECHO_OFF:
		reply[count].resp_retcode = PAM_SUCCESS;
		reply[count].resp = read_string();
		/* PAM frees resp */
		break;
	    case PAM_TEXT_INFO:
		/* ignore it... */
		break;
	    case PAM_ERROR_MSG:
		/* also ignore it... */
		break;
		
	    default:
		/* Must be an error of some sort... */
		free (reply);
		return PAM_CONV_ERR;
	}
    }
    if (reply)
	*resp = reply;
    return PAM_SUCCESS;
}

/*
 * the structure pointing at the conversation function for
 * auth and changing the password
 */
static struct pam_conv pipe_conv = {
     conv_func,
     NULL
};
static struct pam_conv text_conv = {
     misc_conv,
     NULL
};
    
/*
 * A function to process already existing gecos information
 */
static void process_gecos(char *gecos)
{
    char *idx;
    
    if (gecos == NULL)
	return;

    if (!full_name)
	full_name = gecos;    
    idx = strchr(gecos, ',');
    if (idx != NULL) {
	*idx = '\0';
	gecos = idx+1;
    }
    if ((idx == NULL) || (*gecos == '\0')) {
	/* no more fields */
	return;
    }

    if (!office)
	office = gecos;    
    idx = strchr(gecos, ',');
    if (idx != NULL) {
	*idx = '\0';
	gecos = idx+1;
    }

    if (!office_ph)
	office_ph = gecos;
    idx = strchr(gecos, ',');
    if (idx != NULL) {
	*idx = '\0';
	gecos = idx+1;
    }
    if ((idx == NULL) || (*gecos == '\0')) {
	/* no more fields */
	return;
    }
    
    if (!home_ph)
	home_ph = gecos;
    idx = strchr(gecos, ',');
    if (idx != NULL) {
	*idx = '\0';
    }
}

/*
 * invalid_field - insure that a field contains all legal characters
 *
 * The supplied field is scanned for non-printing and other illegal
 * characters.  If any illegal characters are found, invalid_field
 * returns -1.  Zero is returned for success.
 */

int invalid_field(const char *field, const char *illegal)
{
    const char *cp;

    for (cp = field; *cp && isprint (*cp) && ! strchr (illegal, *cp); cp++)
	;
    if (*cp)
	return -1;
    else
	return 0;
}

/*
 * A simple function to compute the gecos field size
 */
static int gecos_size(void)
{
    int len = 0;
    
    if (full_name != NULL)
	len += strlen(full_name);
    if (office != NULL)
	len += strlen(office);
    if (office_ph != NULL)
	len += strlen(office_ph);
    if (home_ph != NULL)
	len += strlen(home_ph);
    return len;
}

/* Snagged straight from the util-linux source... May want to clean up
 * a bit and possibly merge with the code in userinfo that parses to
 * get a list.  -Otto
 *
 *  get_shell_list () -- if the given shell appears in /etc/shells,
 *      return true.  if not, return false.
 *      if the given shell is NULL, /etc/shells is outputted to stdout.
 */
static int get_shell_list(char* shell_name)
{
    FILE *fp;
    int found;
    int len;
    static char buf[1024];

    found = FALSE;
    fp = fopen ("/etc/shells", "r");
    if (! fp) {
        if (! shell_name) printf ("No known shells.\n");
        return FALSE;
    }
    while (fgets (buf, sizeof (buf), fp) != NULL) {
        /* ignore comments */
        if (*buf == '#') continue;
        len = strlen (buf);
        /* strip the ending newline */
        if (buf[len - 1] == '\n') buf[len - 1] = 0;
        /* ignore lines that are too damn long */
        else continue;
        /* check or output the shell */
        if (shell_name) {
            if (! strcmp (shell_name, buf)) {
	        found = TRUE;
                break;
            }
        }
        else printf ("%s\n", buf);
    }
    fclose (fp);
    return found;
}

/* ------- the application itself -------- */
int main(int argc, char *argv[])
{
    int		arg;
    int 	retval;
    char	*progname = NULL;
    pam_handle_t 	*pamh = NULL;
    struct passwd	*pw;
    struct pam_conv     *conv;
     
    /* for lack of a better place to put it... */
    setbuf(stdout, NULL);
    setbuf(stdin, NULL);

    if (geteuid() != 0) {
	fprintf(stderr, "userhelper must be setuid root\n");
	exit(ERR_NO_RIGHTS);
    }

    while (!w_flg && (arg = getopt(argc, argv, "f:o:p:h:cs:tw:")) != -1) {
	/* we process no arguments after -w progname; those are sacred */
	switch (arg) {
	    case 'f':
		f_flg++; full_name = optarg;
		break;
	    case 'o':
		o_flg++; office = optarg;
		break;
	    case 'h':
		h_flg++; home_ph = optarg;
		break;
	    case 'p':
		p_flg++; office_ph = optarg;
		break;
	    case 'c':
		c_flg++;
		break;
	    case 's':
		s_flg++; shell_path = optarg;
		break;
	    case 't':
		t_flg++;
		break;
	    case 'w':
		w_flg++; progname = optarg;
		break;
	    default:
		exit(ERR_INVALID_CALL);
	}
    } 
    /* verify the parameters a little */
#   define SHELL_FLAGS (f_flg || o_flg || p_flg || h_flg || s_flg)
    if ((c_flg && SHELL_FLAGS) || (c_flg && w_flg) || (w_flg && SHELL_FLAGS))
	exit(ERR_INVALID_CALL);

    /* point to the right conversation function */
    conv = t_flg ? &text_conv : &pipe_conv;
    
    /* now try to identify the username we are doing all this work for */
    user_name = getlogin();
    if (user_name == NULL) {
	struct passwd *tmp;
	
	tmp = getpwuid(getuid());
	if (tmp != (struct passwd *)NULL) {
	    user_name = tmp->pw_name;
	    if (user_name != NULL)
		user_name = strdup(user_name);
	    else
		/* weirdo, bail out */
		exit (ERR_UNK_ERROR);
	}
    }
    if ((getuid() == 0) && (argc == optind+1)) {
	/* we have a username supplied on the command line and
	 * the guy running this program is the root */
	user_name = argv[optind];
    }

    /* check for the existance of this user */
    pw = getpwnam(user_name);
    if (pw == (struct passwd *)NULL) {
	/* user not known */
	exit(ERR_NO_USER);
    }

    /* I guess we don't need this anymore */
    endpwent();

    /* okay, start the process */    
    if (c_flg) { /* are we changing the password ? */

	/* Start the PAM session, recommend ourselves as 'passwd' */
	retval = pam_start("passwd", user_name, conv, &pamh);
	if (retval != PAM_SUCCESS)
	    fail_error(retval);

	/* Change the password */
	retval = pam_chauthtok(pamh, 0);
	if (retval != PAM_SUCCESS)
	    fail_error(retval);

    } else if (w_flg) { /* we are a wrapper program */
	/* pick the first existing of /usr/sbin/<progname> and /sbin/<progname>
	 * authenticate <progname>
	 * argv[optind-1] = <progname> (boondoggle unless -- used)
	 * (we know there is at least one open slot in argv for this)
	 * execv(constructed_path, argv+optind);
	 */
	char *constructed_path;

/*sleep(60);*/

	constructed_path = malloc(strlen(progname) + sizeof("/usr/sbin/" + 2));
	if (!constructed_path) exit (ERR_NO_MEMORY);

	strcpy(constructed_path, "/usr/sbin/");
	strcat(constructed_path, progname);
	if (access(constructed_path, X_OK)) {
	    strcpy(constructed_path, "/sbin/");
	    strcat(constructed_path, progname);
	    if (access(constructed_path, X_OK)) {
		exit (ERR_NO_PROGRAM);
	    }
	}

	retval = pam_start(progname, user_name, conv, &pamh);
	if (retval != PAM_SUCCESS)
	    fail_error(retval);

	retval = pam_authenticate(pamh, 0);
	if (retval != PAM_SUCCESS) {
	    pam_end(pamh, retval);
	    fail_error(retval);
	}

	retval = pam_acct_mgmt(pamh, 0);
	if (retval != PAM_SUCCESS) {
	    pam_end(pamh, retval);
	    fail_error(retval);
	}

	/* this is not a session, so do not do session management */

	pam_end(pamh, PAM_SUCCESS);

	/* time for an exec */
	setuid(getuid());
	argv[optind-1] = progname;
	execv(constructed_path, argv+optind-1);
	exit (ERR_EXEC_FAILED);

    } else { /* we are changing some gecos fields */

	char	new_gecos[GECOS_LENGTH+10]; 
	pwdb_type user_unix[2] = { PWDB_UNIX, _PWDB_MAX_TYPES };
	const struct pwdb *_pwdb = NULL;
	const struct pwdb_entry *_pwe = NULL;
	
	/* verify the fields we were passed */
	if (f_flg && invalid_field(full_name, ":,="))
	    exit(ERR_FIELDS_INVALID);
	if (o_flg && invalid_field(office, ":,="))
	    exit(ERR_FIELDS_INVALID);
	if (p_flg && invalid_field(office_ph, ":,="))
	    exit(ERR_FIELDS_INVALID);
	if (h_flg && invalid_field(home_ph, ":,="))
	    exit(ERR_FIELDS_INVALID);
    
	retval = pam_start("chfn", user_name, conv, &pamh);
	if (retval != PAM_SUCCESS)
	    fail_error(retval);

	retval = pam_authenticate(pamh, 0);
	if (retval != PAM_SUCCESS) {
	    pam_end(pamh, retval);
	    fail_error(retval);
	}
	
	retval = pam_acct_mgmt(pamh, 0);
	if (retval != PAM_SUCCESS) {
	    pam_end(pamh, retval);
	    fail_error(retval);
	}
	
	/* Now update the user entry */
	pwdb_start();
	retval = pwdb_locate("user", user_unix, user_name, PWDB_ID_UNKNOWN, &_pwdb);
	if (retval != PWDB_SUCCESS) {
	    pam_end(pamh, PAM_ABORT);
	    fail_error(PAM_ABORT);
	}

	retval = pwdb_get_entry(_pwdb, "gecos", &_pwe);
	if (retval == PWDB_SUCCESS) {
	    /* we have a gecos entry already... process it */
	    process_gecos(_pwe->value);
	}
     
	/* verify that the strigs we get passed are not too long */	
	if (gecos_size() > GECOS_LENGTH) {
	    pwdb_entry_delete(&_pwe);
	    pwdb_delete(&_pwdb);
	    pwdb_end();
	    pam_end(pamh, PAM_ABORT);
	    exit(ERR_FIELDS_INVALID);
	}
	/* Build the new gecos field */
	snprintf(new_gecos, sizeof(new_gecos),
		 "%s,%s,%s,%s",
		 full_name ? full_name : "",
		 office ? office : "",
		 office_ph ? office_ph : "",
		 home_ph ? home_ph : "");

	/* We don't need the pwdb_entry for gecos anymore... */
	pwdb_entry_delete(&_pwe);
	
	retval = pwdb_set_entry(_pwdb, "gecos", new_gecos, 1+strlen(new_gecos),
				NULL, NULL, 0);
	if (retval != PWDB_SUCCESS) {
	    /* try to exit cleanely */
	    pwdb_delete(&_pwdb);
	    pwdb_end();
	    pam_end(pamh, PAM_ABORT);
	    fail_error(PAM_ABORT);
	}

	/* if we change the shell too ... */
	if (s_flg != 0) {

	  /* check that the users current shell is valid... */
	  retval = pwdb_get_entry(_pwdb, "shell", &_pwe);
	  if(!get_shell_list(shell_path) || 
	     !get_shell_list((char*)_pwe->value) ||
	     retval != PWDB_SUCCESS) {
	    fail_error(ERR_SHELL_INVALID);
	    pwdb_entry_delete(&_pwe);
	  }
	    pwdb_entry_delete(&_pwe);

	    retval = pwdb_set_entry(_pwdb, "shell", shell_path, 1+strlen(shell_path),
				    NULL, NULL, 0);
	    if (retval != PWDB_SUCCESS) {
		/* try a clean exit */
		pwdb_delete(&_pwdb);
		pwdb_end();
		pam_end(pamh, PAM_ABORT);
		fail_error(PAM_ABORT);
	    }
	}    
	retval = pwdb_replace("user", _pwdb->source, user_name, PWDB_ID_UNKNOWN, &_pwdb);
	if (retval != PWDB_SUCCESS) {
	    /* clean exit ... */
	    pwdb_delete(&_pwdb);
	    pwdb_end();
	    pam_end(pamh, PAM_ABORT);
	    fail_error(PAM_ABORT);
	}
	pwdb_delete(&_pwdb);
	pwdb_end();	
    }

    /* all done */     
    if (pamh != NULL)
	retval = pam_end(pamh, PAM_SUCCESS);
    if (retval != PAM_SUCCESS)
	fail_error(retval);
    exit (0);
}
